### DISCUSSION TITLE
Node-gyp (through Webpack) attempts to compile C# on Mac

### DISCUSSION THREAD

**Discussion by willianpc** [webpack/webpack#13833] [General]
# Bug report

**What is the current behavior?**
When attempting to bundle a Node.js project that depends on node-gyp, the following error is displayed:
```
ERROR in ./node_modules/node-gyp/lib/Find-VisualStudio.cs 9:6
Module parse failed: Unexpected token (9:6)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
| // This script needs to be compatible with PowerShell v2 to run on Windows 2008R2 and Windows 7.
| 
> using System;
| using System.Text;
| using System.Runtime.InteropServices;
 @ ./node_modules/node-gyp/lib/ sync ^\.\/.*$ ./Find-VisualStudio.cs
 @ ./node_modules/node-gyp/lib/node-gyp.js 41:13-36 190:36-53
```

The same doesn't happen if I simply install the native module (which is also compiled) and run it without Webpack.

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Create a Node.js app that depends on a native module that will be compiled through node-gyp
2. Try to bundle it

**What is the expected behavior?**
As far as I understand, node-gyp should not attempt to compile a C# module in a Mac machine.
I may be wrong or missing something here. My impression is that node-gyp detects the OS somehow to pick the right compiler, but somehow this info may be lost when running through Webpack.
If you guys have any clue, it's much appreciated. If this is not related, please feel free to close the issue and I may try my luck with the node-gyp guys.

Thanks in advance.

**Other relevant information:**
webpack version:  5.45.1
Node.js version: 14.17
Operating System: Mac OS Big Sur 11.4
Additional tools:

**Comment by alexander-akait** [comment]
Can you provide example of the problem, ideally you should ignore these files because we can't parser it, but without configuration hard to say why we try to parse it

**Reply by willianpc** [reply]
Hi, thanks for helping with this.
I made a repo now: https://github.com/willianpc/webpack-node-gyp-case
A simple `npm i` and `npx webpack` should lead you to the issue.
Not sure how familiar you are with node-gyp, but if you run into problems you may need to install all node-gyp dependencies, update xcode, etc.
Feel free to ping me if you stumble into any of these. If do get any of these issues, that is obviously before webpack can runs, so it is most likely unrelated to the issue.

**Reply by alexander-akait** [reply]
this is due:
```
function Gyp () {
  var self = this

  this.devDir = ''
  this.commands = {}

  commands.forEach(function (command) {
    self.commands[command] = function (argv, callback) {
      log.verbose('command', command, argv)
      return require('./' + command)(self, argv, callback)
    }
  })
}
```

`require('./' + command)(self, argv, callback)` is get all modules for current directory and there `Find-VisualStudio.cs` file, you can look at webpack stacktrace and found line there it, do you really need bundle `node-gyp`? Do you need bundle `node_modules`? If not I recommend to use:
```
const nodeExternals = require('webpack-node-externals');
const path = require("path");

module.exports = {
  target: "node",
  mode: "production",
  entry: "./src/index.ts",
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: "ts-loader",
        exclude: /node_modules/,
      },
    ],
  },
  resolve: {
    extensions: [".tsx", ".ts", ".js"],
    fallback: {
      path: require.resolve("path-browserify"),
      util: require.resolve("util/"),
      assert: require.resolve("assert/"),
      https: require.resolve("https-browserify"),
      http: require.resolve("stream-http"),
      buffer: require.resolve("buffer/"),
      url: require.resolve("url/"),
      crypto: require.resolve("crypto-browserify"),
      stream: require.resolve("stream-browserify"),
      zlib: require.resolve("browserify-zlib"),
      os: require.resolve("os-browserify/browser"),
      constants: require.resolve("constants-browserify"),
      timers: require.resolve("timers-browserify"),
      process: require.resolve("process/browser"),
    },
  },
  output: {
    filename: "bundle.js",
    path: path.resolve(__dirname, "lib"),
  },
  externals: [nodeExternals()],
};
```

But in this case I don't know why you set `fallback`, because your `target` is `node` and these modules existing in Node.js.

What is your real target for bundling - `web` or `node`?

**Reply by willianpc** [reply]
Hi Alexander, thanks for taking the time to investigate it.
You are correct, this is a node project and I created this config in a hurry, adding `mode: 'node'` only after I had added the fallbacks. I will update the repo.

We do depend on node-gyp. Basically our API depends on it, and some users of our API are trying to bundle their application, which then triggers this error.

Let me try your suggestion and see how it looks. I come back to you asap.

Cheers,
- Willian

**Reply by willianpc** [reply]
Hi Alexander, I did the changes suggested by you.
I think the ideal solution for any backend application bundled with Webpack is to keep every `node_module` dependency out of the bundle.
Not only because of dynamic requires, but also due to other less common cases, like node-gyp compiling native modules and the binary files from these native modules.
Then I would imagine that a more clear step by step for this would be something like this:

1. Bundle the application without node_modules dependencies by using `node-externals`. This prevents Webpack from bundling them and keep a proper `require('module')` in place
2. Ship the bundle with package.json / package-lock.json /etc
3. During deployment process, run `npm i --production` or one of the similar options in order to get all the external dependencies properly built and to be require by the bundle

What do you think?

I am not sure if this isn't already exposed somewhere, but I wanted to leave this here in case someone else stumbles in a similar scenario.

Thanks one more time!

**Reply by alexander-akait** [reply]
> I think the ideal solution for any backend application bundled with Webpack is to keep every node_module dependency out of the bundle.

We are bundler :smile: `externals` to keep them, different projects use different solutions, 

> What do you think?

Yes, you can use this scenario. 

Anyway you can help us and improve our docs here https://webpack.js.org/guides/author-libraries/,

**Reply by willianpc** [reply]
Right. Well, that was productive, Alexander.
Thank you very much for the insights.
And I will take a look at the doc contribution when I am less busy with work.

Have a nice day.
Cheers!

**Comment by OscarHedeby** [comment]
For anybody else who might get this error, try to see if you by mistake have installed npm as a dependency or simply run npm uninstall npm.
